src/model.py

```diff
@@ -2083,44 +2033,42 @@ def init_components(self, subtasks):
                     out_channels=cnn_dim, 
                     kernel_size=kernel_size
                 ),
-                "attn_linear": self.linear_block(cnn_dim*2, 3)
             })
 
         for layer in range(1, shared_layers):
             # TODO use split_cnn_blocks in all following layers?
-            shared.update({
-                f"cnn_{layer}": self.cnn_block(
+            shared[f"cnn_{layer}"] = self.cnn_block(
                     in_channels=cnn_dim, 
                     out_channels=cnn_dim, 
                     kernel_size=kernel_size
                 )
-            })
-        components.update(torch.nn.ModuleDict({
-            "shared": shared
-        }))
-
-        # similar to RACL setup
-        relations = torch.nn.ModuleDict({
-            f"stack_{i}": torch.nn.ModuleDict({
-                f"{rel[0]}_at_{rel[1]}": self.attn_block(cnn_dim)
-                for rel in attention_relations
-            })
-            for i in range(stack_count)
-        })
-        components.update(torch.nn.ModuleDict({
-            "relations": relations,
-        }))
+        components["shared"] =  shared
 
-        if stack_count > 0:  # when stack == 0 no attention should be made. No task-wise cnn either.. 
-            prev_task = None
-            for task in self.subtasks:
-                for relation in components["relations"]["stack_0"]:
-                    if relation.startswith(task) and (task != prev_task):
-                        components[task]["attn_linear"] = self.linear_block(cnn_dim*2, 3)
-                        prev_task = task
+        ## Relation layers: similar to RACL setup
+        relations = torch.nn.ModuleDict({})
+        attn_linears = torch.nn.ModuleDict({})
+        for stack in range(stack_count):
+            for rel in attention_relations:
+                relations[f"stack_{stack}"] = torch.nn.ModuleDict({
+                    f"{rel[0]}_at_{rel[1]}": torch.nn.ModuleDict({
+                        "attn": self.attn_block(cnn_dim),
+
+                        # new linear for each attn relation for first_task logits using attn information
+                        "linear": self.linear_block(cnn_dim*2, 3),
+                    })
+                })
```

Comment:
Each relation now consists of an attention layer and a linear layer for producing task-wise logits